velox/common/memory/tests/SharedArbitratorTest.cpp

```diff
@@ -1753,6 +1753,72 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, raceBetweenTaskTerminateAndReclaim) {
   Task::testingWaitForAllTasksToBeDeleted();
 }
 
+DEBUG_ONLY_TEST_F(SharedArbitrationTest, raceBetweenMaybeReserveAndTaskAbort) {
+  setupMemory(kMemoryCapacity, 0);
+  const int numVectors = 10;
+  std::vector<RowVectorPtr> vectors;
+  for (int i = 0; i < numVectors; ++i) {
+    vectors.push_back(newVector());
+  }
+  createDuckDbTable(vectors);
+
+  std::shared_ptr<core::QueryCtx> queryCtx = newQueryCtx(kMemoryCapacity);
+  ASSERT_EQ(queryCtx->pool()->capacity(), 0);
+
+  // Create a fake query to hold some memory to trigger memory arbitration.
+  std::shared_ptr<core::QueryCtx> fakeQueryCtx = newQueryCtx(kMemoryCapacity);
+  auto fakeLeafPool = fakeQueryCtx->pool()->addLeafChild("fakeLeaf");
+  Allocation fakeAllocation{
+      fakeLeafPool.get(),
+      fakeLeafPool->allocate(kMemoryCapacity / 3),
+      kMemoryCapacity / 3};
+
+  std::unique_ptr<Allocation> injectAllocation;
+  std::atomic<bool> injectAllocationOnce{true};
+  SCOPED_TESTVALUE_SET(
+      "facebook::velox::common::memory::MemoryPoolImpl::maybeReserve",
+      std::function<void(memory::MemoryPool*)>([&](memory::MemoryPool* pool) {
+        if (!injectAllocationOnce.exchange(false)) {
+          return;
+        }
+        // The injection memory allocation (with the given size) makes sure that
+        // maybeReserve fails and abort this query itself.
+        const size_t injectAllocationSize =
+            pool->freeBytes() + arbitrator_->stats().freeCapacityBytes;
+        injectAllocation.reset(new Allocation{
+            fakeLeafPool.get(),
+            fakeLeafPool->allocate(injectAllocationSize),
+            injectAllocationSize});
+      }));
+
+  const int numDrivers = 1;
+  const auto spillDirectory = exec::test::TempDirectoryPath::create();
+  std::thread queryThread([&]() {
+    VELOX_ASSERT_THROW(
+        AssertQueryBuilder(duckDbQueryRunner_)
```

Comment:
nit: you can skip duckDbQueryRunner_ here and replace assertResults with copyResults()

```
   AssertQueryBuilder()...copyResults(pool());
```